Have EventTarget subclass CanMakeCheckedPtr

cdumez opened the pull request:
#### 621b09e126a65fbc16a33b9c510e4cbb8d438d78
<pre>
Have EventTarget subclass CanMakeCheckedPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=261748">https://bugs.webkit.org/show_bug.cgi?id=261748</a>

Reviewed by Brent Fulgham.

Have EventTarget subclass CanMakeCheckedPtr and start using CheckedPtr in a few
places instead of WeakPtr.

CheckedPtr is more efficient and having EventTarget subclasses (such as Node)
support it will help our smart pointer effort.

This tested as perf-neutral on MotionMark and Membuster. This tested as a
0.4% progression (with 98% confidence) on Speedometer 3.

* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/dom/Document.h:
(WebCore::Document::incrementPtrCount const):
(WebCore::Document::decrementPtrCount const):
* Source/WebCore/dom/EventTarget.cpp:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isDocumentElementRenderer const):
(WebCore::RenderObject::node const):
(WebCore::RenderObject::document const):
(WebCore::RenderObject::treeScopeForSVGReferences const):
(WebCore::RenderObject::nodeForNonAnonymous const):

Canonical link: <a href="https://commits.webkit.org/268146@main">https://commits.webkit.org/268146@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cfdfe9a993a3973a905fec110fc829a020a2fec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19181 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20707 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17634 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19032 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19319 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19064 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19196 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16411 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21589 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16416 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17173 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17445 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17345 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21522 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17938 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17003 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4475 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21370 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -269,7 +269,7 @@ class RenderObject : public CachedImageClient {
     virtual bool isRenderVTTCue() const { return false; }
 
-    bool isDocumentElementRenderer() const { return document().documentElement() == m_node; }
+    bool isDocumentElementRenderer() const { return document().documentElement() == &m_node.get(); }
```
This could be `== m_node.ptr()`, I think?

cdumez replied:
Yes, will fix. I didn't realize we had `ptr()` on CheckedRef.

brentfulgham commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -495,8 +495,7 @@ class RenderObject : public CachedImageClient {
             return nullptr;
-        ASSERT(m_node);
-        return m_node.get(); 
+        return &m_node.get();
```
`m_node.ptr()`?

cdumez replied:
Will fix.

brentfulgham reviewed (commented):
Looks good. Nice that it's also a perf improvement! r=me

cdumez committed:
Have EventTarget subclass CanMakeCheckedPtr
https://bugs.webkit.org/show_bug.cgi?id=261748

Reviewed by Brent Fulgham.

Have EventTarget subclass CanMakeCheckedPtr and start using CheckedPtr in a few
places instead of WeakPtr.

CheckedPtr is more efficient and having EventTarget subclasses (such as Node)
support it will help our smart pointer effort.

This tested as perf-neutral on MotionMark and Membuster. This tested as a
0.4% progression (with 98% confidence) on Speedometer 3.

* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/dom/Document.h:
(WebCore::Document::incrementPtrCount const):
(WebCore::Document::decrementPtrCount const):
* Source/WebCore/dom/EventTarget.cpp:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isDocumentElementRenderer const):
(WebCore::RenderObject::node const):
(WebCore::RenderObject::document const):
(WebCore::RenderObject::treeScopeForSVGReferences const):
(WebCore::RenderObject::nodeForNonAnonymous const):

Canonical link: https://commits.webkit.org/268146@main
